src/native/simplekdf.rs

```diff
@@ -0,0 +1,280 @@
+    }
+}
+
+impl KeyInfo {
+    /// Create a new key info from the given kkey `Object`
+    pub fn new_from_object(obj: &Object) -> Result<KeyInfo> {
+        Ok(KeyInfo {
+            sensitive: obj.is_sensitive(),
+            extractable: obj.is_extractable(),
+            always_sensitive: obj.is_always_sensitive(),
+            never_extractable: obj.is_never_extractable(),
+            value: obj.get_attr_as_bytes(CKA_VALUE)?.clone(),
+        })
+    }
+}
+
+/// Represents a Simple Key Derivation Function (KDF) operation
+///
+/// This handles mechanisms like `CKM_CONCATENATE_*`, and
+/// `CKM_XOR_BASE_AND_DATA`.
+#[derive(Debug)]
+pub struct SimpleKDFOperation {
+    /// Tracks if the derive operation has been completed
+    finalized: bool,
+    /// The specific PKCS#11 Simple KDF mechanism being used
+    mech: CK_MECHANISM_TYPE,
+    /// The additional key handle in CKM_CONCATENATE_BASE_AND_KEY
+    key_handle: Option<[CK_OBJECT_HANDLE; 1]>,
+    /// The additional key in CKM_CONCATENATE_BASE_AND_KEY
+    key: Option<KeyInfo>,
+    /// The additional data to concatenate
+    data: Option<Vec<u8>>,
+}
+
+unsafe impl Send for SimpleKDFOperation {}
+unsafe impl Sync for SimpleKDFOperation {}
+
+impl SimpleKDFOperation {
+    /// Creates a new Simple KDF operation based on the provided mechanism
+    ///
+    /// Dispatches to specific constructors based on the mechanism type.
+    pub fn new(mech: &CK_MECHANISM) -> Result<SimpleKDFOperation> {
+        match mech.mechanism {
+            CKM_CONCATENATE_BASE_AND_KEY => {
+                let object_handle = cast_params!(mech, CK_OBJECT_HANDLE);
+                Ok(SimpleKDFOperation {
+                    finalized: false,
+                    mech: mech.mechanism,
+                    key_handle: Some([object_handle]),
+                    key: None,
+                    data: None,
+                })
+            }
+            CKM_CONCATENATE_BASE_AND_DATA
+            | CKM_CONCATENATE_DATA_AND_BASE
+            | CKM_XOR_BASE_AND_DATA => {
+                let params = cast_params!(mech, CK_KEY_DERIVATION_STRING_DATA);
+
+                let data = bytes_to_vec!(params.pData, params.ulLen);
+                Ok(SimpleKDFOperation {
+                    finalized: false,
+                    mech: mech.mechanism,
+                    key_handle: None,
+                    key: None,
+                    data: Some(data),
+                })
+            }
+            _ => return Err(CKR_MECHANISM_INVALID)?,
+        }
+    }
+}
+
+impl MechOperation for SimpleKDFOperation {
+    fn mechanism(&self) -> Result<CK_MECHANISM_TYPE> {
+        Ok(self.mech)
+    }
+
+    fn finalized(&self) -> bool {
+        self.finalized
+    }
+    #[cfg(feature = "fips")]
+    fn fips_approved(&self) -> Option<bool> {
+        Some(true)
+    }
+    fn requires_objects(&self) -> Result<&[CK_OBJECT_HANDLE]> {
+        if self.mech == CKM_CONCATENATE_BASE_AND_KEY {
```

Comment:
This one is probably more readabe if you return only at the end:
```
{
   if self.mech == CKM_CONCATENATE_BASE_AND_KEY && let Some(h) = &self.key_handle {
      return Ok(h);
   }
   Err(CKR_OK)?
}

Reply:
Or maybe even just `if let Some(h) = &self.key_handle {` only, I mean all other cases already set key_handle to none so it is safe and tidier

Reply:
Done.